File: core-model/src/main/java/com/zenyte/server/AttributesSerializable.java
package com.zenyte.server;

import com.google.gson.Gson;

import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

/**
 * @author Kryeus / John J. Woloszyk
 */
public abstract class AttributesSerializable extends Attributes {

    public static <T extends AttributesSerializable> T getFromFile(
            String file,
            T attributesSerializable) throws IOException {
        final Path path = Path.of(file);
        if (Files.exists(path)) {
            T attributes = new Gson().fromJson(
                    Files.readString(path),
                    attributesSerializable.getType()
            );
            if (attributes == null) {
                return attributesSerializable;
            }
            attributes.setFile(file);
            return attributes;
        } else {
            return attributesSerializable;
        }
    }

    public abstract Type getType();

    private transient String file;

    public AttributesSerializable(String file) {
        this.file = file;
    }

    void setFile(String file) {
        this.file = file;
    }

    public void write() {
        JsonUtil.toJson(this, file);
    }

    public void set(String key, Object value) {
        super.set(key, value);
        write();
    }

    public void remove(Object key) {
        super.remove(key);
        write();
    }

    public void setInt(String key, int set) {
        super.setInt(key, set);
        write();
    }

    public void removeInt(String key) {
        super.removeInt(key);
        write();
    }

    public void removeDouble(String key) {
        super.removeDouble(key);
        write();
    }

    public void setBoolean(String key, boolean set) {
        super.setBoolean(key, set);
        write();
    }

    public void removeBoolean(String key) {
        super.removeBoolean(key);
        write();
    }

    public void setLong(String key, long set) {
        super.setLong(key, set);
        write();
    }

    public void removeLong(String key) {
        super.removeLong(key);
        write();
    }

    public void setString(String key, String set) {
        super.setString(key, set);
        write();
    }

    public void removeString(String key) {
        super.removeString(key);
        write();
    }

    public void setList(String key, List list) {
        super.setList(key, list);
        write();
    }

}

File: core-model/src/main/kotlin/org/jesse/server/AttributesSerializable.kt
package org.jesse.server

import com.google.gson.Gson
import java.io.IOException
import java.lang.reflect.Type
import java.nio.file.Files
import java.nio.file.Path

/**
 * @author Kryeus / John J. Woloszyk
 */
abstract class AttributesSerializable(@field:Transient private var file: String) : Attributes() {

    abstract fun getType(): Type

    internal fun setFile(file: String) {
        this.file = file
    }

    open fun write() {
        JsonUtil.toJson(this, file)
    }

    override fun set(key: String, value: Any?) {
        super.set(key, value)
        write()
    }

    override fun remove(key: Any?) {
        super.remove(key)
        write()
    }

    override fun setInt(key: String, set: Int) {
        super.setInt(key, set)
        write()
    }

    override fun removeInt(key: String) {
        super.removeInt(key)
        write()
    }

    override fun removeDouble(key: String) {
        super.removeDouble(key)
        write()
    }

    override fun setBoolean(key: String, set: Boolean) {
        super.setBoolean(key, set)
        write()
    }

    override fun removeBoolean(key: String) {
        super.removeBoolean(key)
        write()
    }

    override fun setLong(key: String, set: Long) {
        super.setLong(key, set)
        write()
    }

    override fun removeLong(key: String) {
        super.removeLong(key)
        write()
    }

    override fun setString(key: String, set: String?) {
        super.setString(key, set)
        write()
    }

    override fun removeString(key: String) {
        super.removeString(key)
        write()
    }

    override fun setList(key: String, list: List<*>?) {
        super.setList(key, list)
        write()
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun <T : AttributesSerializable> getFromFile(
                file: String,
                attributesSerializable: T): T {
            val path = Path.of(file)
            if (Files.exists(path)) {
                val attributes = Gson().fromJson<T>(
                        Files.readString(path),
                        attributesSerializable.getType()
                )
                if (attributes == null) {
                    return attributesSerializable
                }
                attributes.setFile(file)
                return attributes
            } else {
                return attributesSerializable
            }
        }
    }
}
